rust/storage/src/s3.rs

```
@@ -169,6 +180,118 @@ impl S3Storage {
         }
     }
 
+    pub(super) async fn get_key_ranges(&self, key: &str) -> (i64, Vec<(i64, i64)>) {
+        let part_size = self.download_part_size_bytes as i64;
+        let head_res = self
+            .client
+            .head_object()
+            .bucket(self.bucket.clone())
+            .key(key)
+            .send()
+            .await;
+        let content_length = match head_res {
+            Ok(res) => match res.content_length {
+                Some(len) => len,
+                None => {
+                    panic!("No content length in head response");
+                }
+            },
+            Err(e) => {
+                panic!("Error in head request: {:?}", e);
```

should these be panics or Errs?

---

We should propagate this error!

---

oops